### DISCUSSION TITLE
Mutation through let binding (and no &)?

### DISCUSSION THREAD

**Discussion by dholmin** [hylo-lang/hylo#1522] [Q&A] [answered]
Hi all,

I just started out playing with the compiler (on MacOS), and one of the first things I typed that I didn't expect to compile, did. Here is the code:

```
type Thing : Deinitializable {
  public let position: Int
  public let speed: Int
  public memberwise init
}

fun update_physics(_ t: Thing)
{
  t.position += t.speed
}

public fun main()
{
  let t = Thing(position: 0, speed: 1)
  update_physics(t)
  update_physics(t)
  print(t.position)
}
```

This compiles and prints 2. I was expecting it to complain about mutation through a let binding and missing &s and maybe other things. What am I missing here?

**Comment by dabrahams** [comment, marked as answer]
Nothing; it's a bug!

**Reply by dholmin** [reply]
Glad I haven’t misunderstood everything. :) Maybe I should experiment with Ints instead.

**Reply by dabrahams** [reply]
If `Int`s are different, that's a bug too

**Reply by dholmin** [reply]
What do you mean? Using Ints the compiler enforces more rules, but that should be more correct, right? This is a version that uses an Int instead of a struct:

```
fun update_physics(_ thing: Int)
{
  thing += 1
}

public fun main()
{
  let t = 0 
  update_physics(t)
  update_physics(t)
  print(t)
}
```

Compiling gives me this:

```
main.hylo:3.3-8: error: illegal mutable access
  thing += 1
```

After correcting that and a few more things, I ended up with this, which compiles and runs:

```
fun update_physics(_ thing: inout Int)
{
  thing += 1
}

public fun main()
{
  var t = 0 
  update_physics(&t)
  update_physics(&t)
  print(t)
}
```

It's more in line with what I would expect, but is it correct that the inout `thing` can be mutated without prefixing it with &?

**Reply by dabrahams** [reply]
What I mean is: an `Int` is a type defined in the standard library; it is not substantially different from `Thing`.  If it acts differently, that's a bug.  That said, obviously, `Int` has are no stored properties accessible to user code, so you can't test exactly the same things (unless you pass [`--import-builtin`](https://github.com/hylo-lang/hylo/blob/6476490738421da1007c6ad7d577db1fc5748edb/Docs/CLI.md#--import-builtin), presumably)

**Reply by dholmin** [reply]
I see. Thanks.
